File: src/com/leetcode/p977/Solution.java
package com.leetcode.p977;

public class Solution {

    /**
     * 思路：给的参数是一个升序数组，但开头可能是负数，平方后会比较大。
     *
     * 按绝对值比较，取两头是最大值，逐步向内遍历。
     * @param nums
     * @return
     */
    public static int[] sortedSquares(int[] nums) {
        int start = 0;
        int end = nums.length - 1;
        int index = end;

        int[] result = new int[nums.length];
        while (index >= 0) {
            int startVal = nums[start] * nums[start];
            int endVal = nums[end] * nums[end];
            if (startVal > endVal) {
                result[index] = startVal;
                start++;
            } else {
                result[index] = endVal;
                end--;
            }
            index--;
        }
        return result;
    }
}

File: src/com/leetcode/p977/solution.go
package p977

func sortedSquares(nums []int) []int {
	start, end := 0, len(nums)-1
	index := end
	result := make([]int, len(nums))

	for index >= 0 {
		startVal := nums[start] * nums[start]
		endVal := nums[end] * nums[end]

		if startVal > endVal {
			result[index] = startVal
			start++
		} else {
			result[index] = endVal
			end--
		}
		index--
	}
	return result
}
